packages/statemanager/src/ethersStateManager.ts

```diff
@@ -0,0 +1,384 @@
+    ;(newState as any)._cache = this._cache
+    return newState
+  }
+
+  /**
+   * Sets the new block tag used when querying the provider and clears the
+   * internal cache.
+   * @param blockTag - the new block tag to use when querying the provider
+   */
+  setBlockTag(blockTag: BlockTagType): void {
+    if (typeof blockTag === 'bigint') {
+      this.blockTag = bigIntToHex(blockTag)
+    } else {
+      this.blockTag = blockTag
+    }
+    this.clearCache()
+  }
+
+  /**
+   * Clears the internal cache so all accounts, contract code, and storage slots will
+   * initially be retrieved from the provider
+   */
+  clearCache(): void {
+    this.contractCache.clear()
+    this.storageCache.clear()
+    this._cache.clear()
+  }
+
+  /**
+   * Gets the code corresponding to the provided `address`.
+   * @param address - Address to get the `code` for
+   * @returns {Promise<Buffer>} - Resolves with the code corresponding to the provided address.
+   * Returns an empty `Buffer` if the account has no associated code.
+   */
+  async getContractCode(address: Address): Promise<Buffer> {
+    let codeBuffer = this.contractCache.get(address.toString())
+    if (codeBuffer !== undefined) return codeBuffer
+    const code = await this.provider.getCode(address.toString(), this.blockTag)
+    codeBuffer = toBuffer(code)
+    this.contractCache.set(address.toString(), codeBuffer)
+    return codeBuffer
+  }
+
+  /**
+   * Adds `value` to the state trie as code, and sets `codeHash` on the account
+   * corresponding to `address` to reference this.
+   * @param address - Address of the `account` to add the `code` for
+   * @param value - The value of the `code`
+   */
+  async putContractCode(address: Address, value: Buffer): Promise<void> {
+    // Store contract code in the cache
+    this.contractCache.set(address.toString(), value)
+  }
+
+  /**
+   * Gets the storage value associated with the provided `address` and `key`. This method returns
+   * the shortest representation of the stored value.
+   * @param address - Address of the account to get the storage for
+   * @param key - Key in the account's storage to get the value for. Must be 32 bytes long.
+   * @returns {Buffer} - The storage value for the account
+   * corresponding to the provided address at the provided key.
+   * If this does not exist an empty `Buffer` is returned.
+   */
+  async getContractStorage(address: Address, key: Buffer): Promise<Buffer> {
+    // Check storage slot in cache
+    const accountStorage: Map<string, Buffer> | undefined = this.storageCache.get(
+      address.toString()
+    )
+    let storage: Buffer | string | undefined
+    if (accountStorage !== undefined) {
+      storage = accountStorage.get(key.toString('hex'))
+      if (storage !== undefined) {
+        return storage
+      }
+    }
+
+    // Retrieve storage slot from provider if not found in cache
+    storage = await this.provider.getStorageAt(
```

Comment:
ideally we should also use getProof with storage slots in consistency with how to get account, and verify the proof

Reply:
The only reason I'm using `getProof` in `getAccountFromProvider` is to provide the correct data for the `cache` which needs balance/nonce/codeHash/storageRoot.  I'm not sure I see the value of doing getProof for storage slots.  Just feels like a lot of extra work unless we don't trust the provider and that seems beyond the scope of this code.

Reply:
Proof validation can happen at another layer and not inside the state-manager itself. Could also mention that the state-manager expects trusted data.